crates/matrix-sdk-common/src/js_tracing.rs

```diff
@@ -13,86 +13,207 @@
 // limitations under the License.
 
 //! Utilities for `tracing` in JS environments
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    fmt,
+    fmt::Debug,
+    io,
+    rc::Rc,
+    sync::atomic::{AtomicU32, Ordering},
+};
 
-use std::{fmt, fmt::Write};
-
-use tracing::{
-    field::{Field, Visit},
-    Event, Level, Subscriber,
+use tracing::{level_filters::LevelFilter, Level, Metadata};
+use tracing_subscriber::fmt::{
+    format::{DefaultFields, Format, Pretty},
+    MakeWriter, Subscriber,
 };
-use tracing_subscriber::layer::Context;
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+extern "C" {
+    /// The type of a javascript-side `Logger` object which we can use to log
+    /// from the rust side.
+    pub type JsLogger;
+
+    #[wasm_bindgen(method)]
+    pub fn debug(this: &JsLogger, data: &JsValue);
+
+    #[wasm_bindgen(method)]
+    pub fn info(this: &JsLogger, data: &JsValue);
+
+    #[wasm_bindgen(method)]
+    pub fn warn(this: &JsLogger, data: &JsValue);
+
+    #[wasm_bindgen(method)]
+    pub fn error(this: &JsLogger, data: &JsValue);
+}
 
-/// An implementation of `tracing_subscriber::layer::Layer` which directs all
-/// events to the JS console
-#[derive(Debug, Default)]
-pub struct Layer {}
+/// A [`MakeWriter`] which will construct an [`io::Write`] instance that will
+/// write to either a [`JsLogger`] or the JavaScript console.
+///
+/// You can either construct this as part of an entire [`Subscriber`], via
+/// [`make_tracing_subscriber`], but it's also possible to
+/// call [`MakeJsLogWriter::new`] or [`MakeJsLogWriter::new_with_logger`] and
+/// then feed the result into
+/// [`tracing_subscriber::fmt::SubscriberBuilder::with_writer`], for example:
```

Comment:
Either "either" has to be removed from the sentence, or the "but it's also possible" bit has to be reworded.